CodeModel/CodeParsers/CSharp/CaDETClassMetricCalculator.cs

```diff
@@ -37,10 +37,127 @@ internal class CaDETClassMetricCalculator
+            var usedMethods = invokedMethods.Union(accessedAccessors);
+            var usedInheritedMethods = protectedParentMembers.Select(m => m.Name).Intersect(usedMethods.Select(m => m.Name));
+            return usedInheritedMethods.Count();
+        }
+
+        private static double GetWMCOfNotAccessorOrMuttatorMethods(CaDETClass parsedClass)
+        {
+            return parsedClass.Members.Where(m => !m.Type.Equals(CaDETMemberType.Property)).Sum(m => m.Metrics[CaDETMetric.CYCLO]);
+        }
+
+        // Public attributes for NOPA metric do not include constants and static fields. (Object-oriented metrics in practice)
+        private static int CountPublicAttributes(CaDETClass parsedClass)
+        {
+            var publicFields = GetPublicFields(parsedClass.Fields);
+            var staticFields = (from field in publicFields
+                             from modifier in field.Modifiers
+                             where modifier.Value.Equals(CaDETModifierValue.Static)
+                             select field).ToList();
+            return publicFields.Where(f => !staticFields.Contains(f)).Count();
+        }
+
+        private static double CountWeightOfClass(CaDETClass parsedClass)
+        {
+            return (double)CountFunctionalPublicMethods(parsedClass.Members) / (CountPublicProperties(parsedClass.Members) + GetPublicFields(parsedClass.Fields).Count());
+        }
+
+        // Public fields for WOC metric do not include constants. (Object-oriented metrics in practice)
+        private static IEnumerable<CaDETField> GetPublicFields(List<CaDETField> fields)
+        {
+            var publicFields = FindFieldsWithModifier(fields, CaDETModifierValue.Public);
+            var constants = FindFieldsWithModifier(publicFields, CaDETModifierValue.Const);
+            return publicFields.Where(f => !constants.Contains(f));
+        }
+
+        private static IEnumerable<CaDETField> FindFieldsWithModifier(IEnumerable<CaDETField> fields, CaDETModifierValue modifierValue)
+        {
+            return (from field in fields
+                    from modifier in field.Modifiers
+                    where modifier.Value.Equals(modifierValue)
+                    select field);
+        }
+
+        private static int CountPublicProperties(IEnumerable<CaDETMember> members)
+        {
+            var properties = members.Where(m => m.Type.Equals(CaDETMemberType.Property));
+            var publicProperties = FindMembersWithModifier(properties, CaDETModifierValue.Public);
+            return publicProperties.Count();
+        }
+
+        // Functional public methods do not include get/set, constructor and abstract methods. (Object-oriented metrics in practice)
+        private static int CountFunctionalPublicMethods(List<CaDETMember> members)
+        {
+            var methods = GetMethods(members);
+            var publicMethods = FindMembersWithModifier(methods, CaDETModifierValue.Public);
```

Comment:
Maybe not for this PR, but something to keep in mind - maybe the GetMethods method of CaDETClass can have optional parameters - CaDETModifierValue withModifier, CaDETModifierValue withoutModifier (or maybe even lists of these).

We will see how new metrics utilize modifiers and refactor accordingly.